SemanticDB Specification, Version 3.0.0

xeno-by opened the pull request:
## Motivation

It's been an exciting year for SemanticDB 🎉 

In the beginning of 2017, we created [semanticdb-scalac](http://scalameta.org/tutorial/#semanticdb-scalac) that was then immediately picked up by [Scalafix](https://scalacenter.github.io/scalafix/). June 2017 was the start of active development of [Metadoc](scalameta.org/metadoc). In November 2017, we welcomed [scalameta/language-server](https://github.com/scalameta/metals) (which is now called Metals, by the way 🤘). Finally, just a week ago we started working on [SemanticDB support for Rsc](https://github.com/xeno-by/reasonable-scala/commits/10).

Our experience with SemanticDB shows that it's not just useful for toy projects, but that it can scale to huge codebases that span millions of lines of code. Our experiments with indexing SemanticDB data into relational databases (https://github.com/scalameta/scalameta/pull/1174 and https://github.com/scalameta/metals/pull/94) demonstrate that resulting databases have reasonable size and single/double-digit millisecond response times.

So far, we've been more or less flying under the radar. Our protobuf schema lives in `package org.langmeta.internal.semanticdb.schema`, and our implementation is our specification. However, now that we've seen evidence that SemanticDB is a good idea, let's make things more official.

## Proposal

  * Let's write an official specification for SemanticDB and commit to having the specification drive the implementations.
  * Let's commit to proper versioning of the SemanticDB data model.
  * Let's extract the protobuf schema into a dedicated project.
  * Let's put the corresponding data structures into a public package.

## Implementation

In this pullrequest, I'd like to present one approach to implementing this proposal that @olafurpg and I have been working on this week. In a nutshell:
  * A specification document was added with the intention to document all the little implementation details that weren't properly communicated until now.
  * The data structures were refactored to accommodate your feedback and share as much vocabulary with Language Server Protocol as possible. After the schema becomes public, making breaking changes will be very complicated, so we decided to use this occasion to do some cleanup.
  * Thanks to the [recent breakthrough by the authors of ScalaPB](https://github.com/scalapb/ScalaPB/issues/387) and to the fact that our newly created project is so small, we've been able to crosscompile the new SemanticDB infrastructure to all three supported platforms: JVM, JS and Native.
  * The SemanticDB protobuf schema was moved to `org.scalameta %% semanticdb3` under the `scala.meta.semanticdb3` package.
  * All work is being done in a new sbt project that lives in a new package and publishes to a new artifact, which means that existing code that works now will continue working in the near future, and we'll be able to go through a civilized deprecation cycle. It may even be possible to support both the old and the new SemanticDB schemas in both the old and the new versions of Scalameta, but that remains to be seen.

There is still work to be done - some of it explicitly marked as TODO in the specification and some of it lies deep in the internals of `semanticdb-scalac` (with https://github.com/scalameta/scalameta/pull/1218 being the first step) - but it seems that we already have enough substance to have a productive discussion. Let us know what you think!

olafurpg commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+SemanticDB payloads must include the descriptor of the data model that was used
+to produce them in the `model` field. The following descriptors are supported:
+
+<table>
```
Why write manual table instead of markdown table?

xeno-by replied:
I find tables easier to maintain than the markdown ASCII art, which doesn't give control of column widths and breaks down on non-trivial formatting. If you'd like, I can try to migrate to the Markdown style.

olafurpg replied:
Interesting, I agree maintaining markdown tables is a drag. Sure, let's do this 👍

olafurpg commented on semanticdb3/semanticdb3.md:
````diff
@@ -0,0 +1,491 @@
+
+```protobuf
+message Range {
+  int32 start_line = 1;
````
I still have doubts about this change, I believe start/end offsets are a superior format and I have concerns about performance implications. For tooling you will always need to have the buffer available anyways, in which case converting between the two formats is easy.

olafurpg replied:
Fair enough, Index-While-Building seems to use line/column as well https://github.com/apple/swift-clang/blob/stable/test/Index/Store/Inputs/json.c.json

> Also, why does one always need to have the buffer available anyways?

Because you need the ability to account for minor changes in the file contents. Blindly returning reference positions may result in incorrect rename refactorings or bogus highlighted code regions. If you have something like sha256 it's better, but then a single character insert makes the entire snapshot unusable.

olafurpg replied:
the readability of the raw data is improved with line/column, but readability will go down the drain anyways if we use symbol lookup tables to compact the data.

olafurpg replied:
Metals btw already has support to recover from slightly stale semanticdb documents using token edit-distance, and that requires the full text contents. In this space, I think we are actually ahead of alternative technologies and the lesson learned is that you need the text contents anyways so you might as well use start/end.

olafurpg replied:
Regardless, I'm fine with using line/character for the simplicity of manually constructing and validating raw data. I would personally prefer `column` over `character` since I think it's confusing to use `character` like is done in LSP. In my editor and in clang they use `col` 

![screen shot 2018-01-19 at 19 13 53](https://user-images.githubusercontent.com/1408093/35165018-f99acafe-fd4c-11e7-80f9-5832eb26e285.png)

xeno-by replied:
Speaking of always having the document text, what about SQL? When you're working with the SQL index, you don't have the luxury of positions on the fly. I suspect that many large-scale code analysis/rewrite jobs will be similarly restrictive.

I think a combination of a fingerprint and optional client smartness (like the one that is implemented in Metals) is the way to go, but let's defer this discussion to a follow-up update to the spec. There are multiple questions in play here - do we fingerprint just the contents of the source file, or its dependencies as well? do we allow fingerprinting to be buildtool-specific? - and they all seem orthogonal to the goal of this pullrequest.

xeno-by replied:
Speaking of the choice between offset and line/column, let's put a hold on this until we have the benchmark data. Then we will be able to talk about readability, too.

ShaneDelmore replied:
I have noticed the usability of raw offsets isn’t great and will admit to pasting hunks of code in an online character counter to quickly figure out what offset I wanted 😊 
On the other hand, offsets are great for dbs that support interval types, especially for fast overlaps lookups.  We could use a datatype that allows encoding two ints in one value, say a decimal type.

xeno-by replied:
Here's benchmark results: https://github.com/xeno-by/scalameta/commit/be54bca11542580f76b670870318ac238f997663. Looks like the effect of refactoring positions is within the fluctuation window.

olafurpg commented on semanticdb3/semanticdb3.md:
````diff
@@ -0,0 +1,491 @@
+### Identifier
+
+```protobuf
+message Identifier {
````
One issue I have with `Identifier` is that it doesn't specify **what** it's identifying. In LSP there is `TextDocumentIdentifier`, BSP has `BuildTargetIdentifier` while here we have an unqualified `Identifier`. What about calling this `SymbolIdentifier` instead? That would make the relationship between `SymbolIdentifier` and `SymbolInformation` clearer IMO.

olafurpg replied:
After more thinking I don't like `SymbolIdentifier` since that intuitively means "the ID for this symbol". `SymbolOccurrence` is the most accurate and self-explanatory name I can think of, and it would match with clangs IndexWhileBuilding usage of "Occurrence".

xeno-by replied:
That's a valid argument. In PL, `Identifier` is unambiguous, but within a larger context this name may probably be not so clear. Let's think about alternatives.

What I like about `Identifier` is that it promotes a text-centric view of the program, where we aren't talking compiler internals, but are appealing to what is written in the source file. In that sense, I liked `Identifier` and `Definition`, because that's what's written in source code. 

`SymbolOccurrence` and `SymbolInformation` are pushing us away from that model. For example, with a text-centric view, you can have an identifier or a definition without any attached symbol, e.g. created by a parser. However, if you call them symbol occurrence or information, then it kinda doesn't make sense to have these data structures without semantic information. Now, this is not something that we have been doing with SemanticDB yet, but this illustrates the idea.

olafurpg replied:
`SymbolOccurrence` seems to be a thing in swift-clang btw, https://github.com/apple/swift-clang/blob/stable/include/clang/Tooling/Refactor/SymbolOccurrenceFinder.h

My favorite remains `SymbolOccurrence`, and I think it's the most consistent name to pair up with `SymbolInformation`

xeno-by replied:
Let me think about this over the weekend.

olafurpg commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+about [Symbols](#symbol) that are defined in the underlying snippet of code.
+In a sense, this section is analogous to symbol tables [\[21\]][21] in compiler.
+
+`Definition` contains assorted metadata for a `symbol`, as explained below.
```
SymbolInformation

xeno-by replied:
Oops.

MasseGuillaume commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+
+"Identifiers" is a section of a [TextDocument](#textdocument) that represents
+the results of name resolution for identifiers in the underlying snippet of code.
+
```
I would give an example of Identifier/role here.

xeno-by replied:
Will do.

olafurpg commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+Nowadays, there is a clear trend towards standards for communication between developer tools.
+Language Server Protocol (LSP) [\[2\]][2], a protocol that connects programming
+language implementations and text editors, has gained strong industrial support
+and at the time of writing has implementations for over 25 programming languages.
```
and most major editors. I don't think there is any editor with >2% marketshare that doesn't have some LSP client implementation, maybe nano but that still doesn't disqualify "most major editors"

xeno-by replied:
I'll think about diplomatic formulation that does LSP justice.

MasseGuillaume commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+    <td><b>Explanation</b></td>
+  </tr>
+  <tr>
+    <td><code>1 << 1</code></td>
```
I don't understand this part. This is a range? `[1, 1]`

MasseGuillaume commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+  </tr>
+  <tr>
+    <td><code>1 << 1</code></td>
+    <td>Reference</td>
```
How is `1 << 1` Related to reference?

MasseGuillaume replied:
ahhh, It's a bit shift. I would just put 1, 2, 4, ...

olafurpg replied:
The idea is that a symbol occurrence can have multiple roles. Currently, we only have reference/definition but other examples that come to mind include `extends`. See https://youtu.be/jGJhnIT-D2M?t=10m12s how they use this role information in xcode https://youtu.be/jGJhnIT-D2M?t=10m12s

olafurpg replied:
Another potential usecase for role would be to label 

```scala
package scala {
  val Left = scala.util.Left
}
```
as an `alias` so that it's easy to "Find references" for `scala.util.Left` and include results for `scala.Left`

xeno-by replied:
I like the alias idea more and more every day.

olafurpg commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+Mention that beyond protobuf files, there are many ways to use SemanticDB as an
+interchange format, including passing objects around in memory, communication via JSON, etc etc.
+
+## Model
```
Data Model

xeno-by replied:
I thought about changing `Model` to `Data Model` and `Schemas` to `Data Schemas`, but didn't find that the added clarity is worth the added clutter. What's your take on this?

olafurpg replied:
I don't think it adds clutter, jumping straight into the document without context it's not clear what "Model" means.

xeno-by replied:
Okay, I'll update.

ShaneDelmore replied:
FWIW I like “data model” and schema, but not “data schema”

xeno-by commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+<table>
+  <tr>
+    <td width="100px"><b>Value</b></td>
+    <td><b>Explanation</b></td>
```
Good idea 👍

xeno-by commented on semanticdb3/semanticdb3.md:
```diff
@@ -0,0 +1,491 @@
+}
+
+message TextDocument {
+  string model = 255;
```
More flexibility, which I believe is essential. For example, semver uses strings.

xeno-by commented:
Thanks for the comments, folks - I left some replies above and will work on addressing your feedback in the coming days. As for performance, let's discuss it in https://github.com/scalameta/scalameta/pull/1218#issuecomment-359019052.

xeno-by commented:
Very well, looks like we have no objections to this proposal 🎉 

I'll execute on the proposal as follows:
  * [X] Factor out SemanticDB v2 (https://github.com/scalameta/scalameta/pull/1228)
  * [ ] Migrate semanticdb-scalac to the low-level SemanticDB API
  * [ ] Introduce SemanticDB v3 addressing the comments here

In the meanwhile, this pull request will be closed. Thank you, everyone, for your extremely helpful feedback 🤝

xeno-by commented:
I would also like to ask for a moratorium on changes to the SemanticDB schema and the semanticdb-scalac compiler plugin until I finish the migration to v3. /cc @olafurpg @MasseGuillaume @gabro @laughedelic
